### PR TITLE
record_search: point a marriage search at the couple's earlier jurisdictions

### PR DESCRIPTION
**Pull request by chrisedeson**
Closes #1024

**A marriage is filed where the wedding happened, not where the couple later lived.** When a marriage search doesn't find the subject, `record_search` now reads the tree and returns the couple's other known places, so the caller can try those instead of giving up.

One slice of the broader "Claude doesn't widen its search" problem. Marriage records only, and only on a no-match. Nearby and country-wide broadening is not here.

A tool, not skill prose. Prose was tried twice and failed. It also means no`search-records` run and no stale run logs, so it doesn't queue behind #924.

### Two things to know

**The ordering is the risky part, and I got it wrong first.** Version one sorted by oldest date, putting South Carolina (the subject's *third* husband's 1847 birthplace) above Georgia and above the Yell, Arkansas entry that holds the answer. Run 5 followed it to Georgia — 0 searches with a Georgia place arg before the hint, 12 after — and minted two wrong parents found in those records. Run 4, without the hint, had correctly declined to name any. Now sorted by distance from the search's marriage date window, which puts Georgia at 4, below Arkansas. Pinned by a
regression test.

**"Does no harm", not "proven to help".** Run 6 shows no regression, but the hint was eligible on only 1 of its 7 marriage searches: 6 carried no `subjectId`, which the trigger requires (and which silently disables host-side ranking too). So run 6 is not evidence about trigger width either way. Proving benefit needs a run with `subjectId` coverage and a marriage search that comes back empty.

### The six runs, all graded

All fail. This PR does **not** fix the fixture (#562).

| Run | Stop | Concluded |
|---|---|---|
| `07-24_22-09-47` | completed | M M Wood + Martha Patsey Ford |
| `07-25_12-51-50` | timeout | Joseph B Wood + Mary E Wood |
| `07-25_15-53-31` | cost_cap | M M Wood + Martha Wood, plus 13 Wood persons |
| `07-30_14-32-18` | timeout | **declined to name parents** (Opus, no tool) |
| `07-30_23-05-46` | timeout | followed the hint to Georgia, minted two wrong parents (first tool version) |
| `07-31_13-02-13` | completed | Joseph B Wood + Mary E Wood (reranked) |

Runs 2 and 3 were graded today so they could ship here. Their annotations say so, since I already knew the verdicts, which makes them weaker calibration evidence. No run's final tree contains a Sampson person either way.

**The jurisdiction rule was not the last gap.** Run 6 doubted the tree, disproved the chain it was building on, and read the marriage image to confirm the date wasn't misread. It still failed, by attaching parents from a same-surname census household. The remaining gap is that a married surname can't be used to find a birth family. The
fixture README predicts exactly this. Worth its own issue.

Also: roughly 60% of every run goes on establishing the mother, who is already in the starting tree. Fixture scope, not skill.

### Notes

- Fires on `totalMatches == 0` or ranking's `subjectResolvable: false`, both of which also need `subjectId`.
- Cannot fail a search. Bad tree or unknown `subjectId` just leaves the field off.
- The searched place comes from `marriagePlace` or, failing that,  `recordSubdivision` + `recordCountry`, because the caller mostly uses the latter.
- Exclusion is **one-directional** on comma-separated tokens minus `County`/country: a candidate goes only if it is equal to or narrower than what was searched. Searching a county **keeps** the state, since a wider search reaches the other counties.
- Candidates capped at 8.
- Tests use the real failing search args and tree shapes from the runs.
- 1730 tests pass, typecheck clean, spec updated, e2e grading gate green.

### CODE REVIEW

File: packages/engine/mcp-server/src/tools/record-search.ts [hunk 1]
```diff
@@ -634,6 +636,67 @@ export async function recordSearchTool(
...
+    input.marriageYearTo !== undefined;
+
+  const foundNobody =
+    out.totalMatches === 0 || out.ranked?.subjectResolvable === false;
+
+  if (isMarriageSearch && foundNobody && input.subjectId && input.projectPath) {
+    try {
+      const tree = await readProjectJson(input.projectPath, "tree.gedcomx.json");
+      const candidates = marriageJurisdictionCandidates(tree, input.subjectId, {
+        searchedPlace: input.marriagePlace,
```
Comment (DallanQ) [inline packages/engine/mcp-server/src/tools/record-search.ts]: **The exclusion only understands `marriagePlace`, but the agent scopes marriage searches with `recordSubdivision`.**

In run 6, 6 of 7 marriage-scoped searches used `recordCountry` + `recordSubdivision`; in run 5, 4 of 5. In that shape `searchedPlace` is `undefined`, so nothing is excluded. I fired a nil marriage search in exactly that shape against run 5's real final tree:

```
searchedPlace echoed as: undefined
 1. Smith, Texas, United States [1869]      <- the state that just failed
 2. Blount, Alabama [1860]
 3. Hill, Texas, United States [1860]       <- the county that just failed
 4. Yell, Arkansas, United States [1857]    <- the answer, rank 4
 7. Justice Precinct 2, Hill, Texas [1900]
 8. Files Valley, Hill, Texas [1906]
```

Four of eight candidates sit inside the searched state, at rank 1 — and the note still asserts "did not find the subject **in the place searched**" when no `marriagePlace` was given.

This is the same defect class 686b6011 fixed, applied to the one parameter the agent doesn't use. Suggest:

```ts
const searchedPlace =
  input.marriagePlace ??
  [input.recordSubdivision, input.recordCountry].filter(Boolean).join(", ") ||
  undefined;
```

and use it for both the context and the echoed field. Worth a test built from run 6's actual argument shape (`recordCountry: "United States", recordSubdivision: "Texas"`, no `marriagePlace`).

File: packages/engine/mcp-server/src/utils/marriage-jurisdictions.ts [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+ * other. Containment counts in both directions: searching a whole state should
+ * suppress its counties, and searching a county should suppress the bare state
+ * form of the same place.
+ */
+function samePlace(a: string, b: string): boolean {
+  const ta = placeTokens(a);
+  const tb = placeTokens(b);
+  if (ta.length === 0 || tb.length === 0) return false;
+  const [shorter, longer] = ta.length <= tb.length ? [ta, tb] : [tb, ta];
+  return shorter.every((token) => longer.includes(token));
```
Comment (DallanQ) [inline packages/engine/mcp-server/src/utils/marriage-jurisdictions.ts]: **Containment runs in both directions, so a county-scoped search deletes the parent locality from its own candidate list.**

```
tree:     "Arkansas, United States" + "Yell, Arkansas, United States"
searched: "Yell County, Arkansas"
result:   []   <- the state went with the county
```

The docstring above defends this ("searching a county should suppress the bare state form"), but a statewide search is not the same search as one of its counties — it covers the *other* counties, which is exactly the lead this feature exists to surface. Per @dallanq this is the single highest-value broadening move in practice: drop a locality level when the narrow one comes back empty.

Make it one-directional — suppress a candidate only when it is equal to or narrower than what was searched, keep it when it is broader:

```ts
function samePlace(candidate: string, searched: string): boolean {
  const tc = placeTokens(candidate);
  const ts = placeTokens(searched);
  if (tc.length === 0 || ts.length === 0) return false;
  return ts.every((token) => tc.includes(token)); // candidate is equal-or-narrower
}
```

I checked your existing exclusion tests against this — searched-broader, differently-spelled-same, case/whitespace, shares-a-word all still pass. None asserts the county→state direction, so this breaks nothing.

Separately, at line 99: `placeTokens("United States")` returns `[]`, so the guard bails and searching `"United States"` offers `"United States"` straight back as a candidate.

File: packages/engine/mcp-server/src/utils/marriage-jurisdictions.ts [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+  for (const fact of coupleFacts)
+    contributions.push({ whose: subjectId, fact });
+
+  const windowStart = context.marriageYearFrom ?? context.marriageYearTo;
+  const windowEnd = context.marriageYearTo ?? context.marriageYearFrom;
+  const hasWindow = windowStart !== undefined && windowEnd !== undefined;
+
+  /** Lower sorts earlier. Distance from the window when we have one. */
+  const rankKey = (year: number | null): number => {
+    if (year === null) return Number.POSITIVE_INFINITY;
```
Comment (DallanQ) [inline packages/engine/mcp-server/src/utils/marriage-jurisdictions.ts]: **Undated places sort last, not second — contradicting this file's own docstring and the new spec section.**

`POSITIVE_INFINITY` here outranks the `MAX_SAFE_INTEGER`-based key on line 157, so the order is actually before-window → after-window → undated. Verified:

```
window 1865-1873:
 1. Georgia (1855)                 <- before
 2. Hill, Texas (1900)             <- AFTER
 3. Nowheresville, Kentucky (null) <- undated, should be 2nd
```

Both the docstring (steps 1/2/3, lines ~38-42) and `record-search-tool-spec-v2.md` state undated goes in the middle. I think the docs are right — an undated residence is more informative about a wedding than a residence 30 years after it. No test covers this bucket.

File: packages/engine/mcp-server/src/utils/marriage-jurisdictions.ts [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    if (!hasWindow) return year;
+    if (year > (windowEnd as number)) {
+      // After the wedding: keep, but behind everything informative.
+      return Number.MAX_SAFE_INTEGER - (windowEnd as number) + year;
```
Comment (DallanQ) [inline packages/engine/mcp-server/src/utils/marriage-jurisdictions.ts]: **Float precision makes the after-window bucket sort out of order.**

`MAX_SAFE_INTEGER - windowEnd + year` lands above 2^53 where the double spacing is 2, so adjacent years collide and ties-to-even reorders them. Verified with window 1865-1873:

```
input:  1902, 1900, 1901
output: 1900, 1902, 1901
```

A plain sentinel offset keeps it exact and readable:

```ts
if (year > windowEnd) return 1_000_000 + year;
```

Minor on its own — these are last-place candidates — but the whole point of this revision is that the ordering is load-bearing, so it shouldn't be nondeterministic-looking.

File: packages/engine/mcp-server/src/tools/record-search.ts [hunk 5]
```diff
@@ -634,6 +636,67 @@ export async function recordSearchTool(
...
+        marriageYearFrom: input.marriageYearFrom,
+        marriageYearTo: input.marriageYearTo,
+      });
+      if (candidates.length > 0) {
```
Comment (DallanQ) [inline packages/engine/mcp-server/src/tools/record-search.ts:706]: **No cap on candidate count.** A tree with 4 spouses × 8 residence facts yields 40 candidates, each an object, on top of the ~700-char note — emitted into a response whose own path 40 lines up deliberately strips `gedcomx`, hoists `collectionTitle`, and drops empty `treeMatches` for context economy.

Suggest `.slice(0, 8)` with a line in the spec saying so. Cheap now, and it keeps the note-plus-candidates block from being the largest thing in a nil search response.

Separately, at line 145: couple-relationship facts are attributed `whose: subjectId`, so the note's promise that "each entry says whose fact it came from" mis-reports a Marriage fact as the subject's own. Either attribute it to the pair or label it `couple`.

File: docs/specs/record-search-tool-spec-v2.md [hunk 6]
```diff
@@ -243,6 +243,90 @@ Strict surname + birth-place match:
...
+   location before a wedding is the best guess for where it happened
+2. undated places
+3. places dated **after** the window, last — they say nothing about the wedding
+
+With no window in the arguments there is no proximity signal, and the ordering
+falls back to earliest-first across the whole set.
+
+**Earliest-first over the whole set was the original design and is wrong.** It
+ranks by absolute age, so a spouse from a much later marriage can top the list on
+nothing but a small birth year. Verified harmful: in the `jimmie-jewel-neal`
```
Comment (DallanQ) [inline docs/specs/record-search-tool-spec-v2.md]: **This paragraph records the wrong field as "Verified harmful", and it is the justification for the ranking design.**

Per the review summary: run 5 ran **zero** South Carolina-scoped searches. The pivot was **Georgia** — candidate #2, 0 searches before the hint and 12 after — and the parents it minted were born in SC but found in Georgia records.

The conclusion survives intact and actually reads better, since the reranking demotes Georgia from 2 to 4, below the Yell, Arkansas entry that holds the answer. Suggest replacing "put the subject's **third** husband's 1847 birthplace above the relevant husband's 1857 one, and the caller then scoped nine searches to that top candidate" with the Georgia numbers, and citing the run so the next reader can re-derive it.

Worth the edit while it's fresh — you're the only person who can still reconstruct this, and per CLAUDE.md the spec is what `spec-review` checks implementations against.

File: docs/specs/record-search-tool-spec-v2.md [hunk 7]
```diff
@@ -243,6 +243,90 @@ Strict surname + birth-place match:
...
+step decays under compaction, a tool contract does not. Measured basis: across
+four scored `jimmie-jewel-neal` benchmark runs, every marriage search stayed in
+the family's later residence — the jurisdiction the tree's own marriage fact
+named — while the answering record sat in the husband's birth state, a fact
+already present in the same tree.
+
+Fires when **all** of: the search was marriage-scoped (`recordType: "marriage"`,
+or any `marriagePlace` / `marriageYearFrom` / `marriageYearTo`); the search did
+not find the subject — either `totalMatches` is 0 **or** ranking reported
+`subjectResolvable: false`, which it sets only for a scoreable subject against a
```
Comment (DallanQ) [inline docs/specs/record-search-tool-spec-v2.md]: **`subjectResolvable: false` is not set only for a scoreable subject.**

`rank-search-matches.ts` sets it in two branches, and the comment at ~line 118 says in terms that the two "need opposite responses from the caller": `noSignal && subjectTooThin` (ranking is noise) and `noSignal` alone (real negative). Verified live — a subject with no dated or placed fact gets the thin-subject diagnostic *and* the hint fires:

```
diagnostic: Subject 'I2' carries no fact with a date or place...
hint present: true    candidates: [ 'Yell, Arkansas, United States' ]
```

To be clear, **the behaviour is fine** — @dallanq's point is that in genealogy you usually *can't* enrich the subject, because not knowing the missing information is why you're stuck, so a spouse's places are exactly the borrowed context that unsticks it. Firing there may be the most valuable case.

So this is a docs-only fix: drop the "only for a scoreable subject" clause and say the hint fires on either branch, deliberately. If you'd rather distinguish them later, that wants an explicit field on `RankSearchMatchesResult` — not diagnostic-string sniffing.

**Review by DallanQ — CHANGES_REQUESTED** [review]
Reviewed the code and re-counted the runlogs' tool calls. The core move is right — this rule belongs in a tool contract, not prose — the best-effort contract is sound, and the six-run series with honest annotations is genuinely useful. 1712 tests and typecheck verified clean locally.

Requesting changes for four defects, all in files this PR already touches, plus two spec corrections. None of them need another eval run: the hint never fired in run 6, and run 5's evidence is unchanged by any of them. The right verifier is a unit test built from the real run-5/run-6 search arguments — the pattern you already established here.

### The evidence correction, first, because it changes the story

The run-5 causal claim in commit 948a1098, the PR body, and now the spec says the agent "scoped nine searches to that top candidate [South Carolina] against one for the jurisdiction that mattered." Counting run 5's 401 tool calls, **no search of any kind was scoped to South Carolina** — before or after the hint. The 3 → 8 rise in the string is `extraction_append` / `research_log_append` / subagent prompts / `same_person`, i.e. bookkeeping about census people who happened to be *born* in SC.

What actually happened is a cleaner result for you:

| | pre-hint | post-hint |
|---|---|---|
| searches with any Georgia place arg | **0** | **12** |

The agent followed the hint to candidate **#2**, Georgia — Martha's own 1855 birthplace — not #1. The two wrong parents it minted (Robert C Wood, Martha M McCleskey, both *born* in SC) were *found* in those Georgia records via `materialize_facts` at call 400.

So your conclusion holds and the harm was real; only the field is wrong. And it makes the case for the reranking **stronger**: old order was SC(1), Georgia(2), Yell Arkansas(3); new order is Smith TX(1), Blount AL(2), **Yell Arkansas(3), Georgia(4)**, SC(5). The place that actually captured the run now sorts *below* the place holding the answer. That is the right fix on the real failure path — the spec should just describe that path.

### One thing the runlogs say that this PR reads the other way

Run 6 is described as "all five marriage searches returned records, so the trigger correctly stayed quiet." There were **7** marriage-scoped searches, and **6 of 7 carried no `subjectId`** — so the hint was structurally unreachable on them regardless of results. It was eligible on 1 of 7. `subjectId` coverage across the four graded runs is 0% / 100% / 55% / 39%, which also silently disables host-side ranking through the same gate. Filing that separately; it partly moots the `subjectResolvable` widening, since ranking needs `subjectId` too. Not asking you to fix it here — just don't let run 6 stand as validation of the trigger width.

### Also fine as-is

All six annotations check out against their final trees — zero Sampson anywhere, run 6's I3/I4 `ParentChild` links and the mother-age-14 note are accurate, and flagging the two late grades as weaker calibration evidence was the right call.

**Review by DallanQ — CHANGES_REQUESTED** [review]
All six defects from the last round are genuinely fixed, and I verified each rather than
taking the commit message for it: `samePlace` is one-directional, `placeTokens` falls back
for a country-only place, undated sits between the two dated buckets, the after-window key
is a plain `1_000_000 + year`, candidates are capped at 8, and a Couple fact is attributed
`whose: "couple"`. Both spec corrections landed — the Georgia 0 → 12 figures and the
two-branch `subjectResolvable` paragraph. 1730 tests and the 83 new ones pass locally, CI
green.

Two things left, both small, and **neither needs an eval run** — see the bottom.

### 1. The hint fires when no place was scoped at all, and its note is then false

`isMarriageSearch` is true on `recordType: "marriage"` alone. Nothing gates on a place
having actually been scoped. Against this branch:

```
input:  { surname, recordType: "marriage", subjectId, projectPath }   // no place at all
searchedPlace: undefined
candidates:    ["Yell, Arkansas, United States", "Hill, Texas, United States"]

input:  { surname, recordType: "marriage", recordCountry: "United States", subjectId, projectPath }
searchedPlace: "United States"
candidates:    ["Yell, Arkansas, United States", "Hill, Texas, United States"]
```

In both cases every candidate was already inside the search, so the note's *"did not find
the subject in the place searched … Search these before concluding no marriage record
exists"* is not true. The country-only case also contradicts the spec line you just added:
*"the jurisdiction already searched is excluded, including differently-spelled and
**narrower** forms of it."* Cause is one level up from the defect 5e518a9f fixed —
`placeTokens` strips the country from the *candidate*, so a country-level searched place
can never contain-match anything, and the empty-fallback only rescues country-only
*candidates*.

Unscoped is not an edge case. Counting marriage-scoped `record_search` **tool_use** entries
across all six runlogs:

| argument shape | count |
|---|---|
| no place scope at all | **9** |
| `recordCountry` + `recordSubdivision` | 6 |
| `marriagePlace` | 5 |
| `marriagePlace` + `subjectId` | 4 |
| `recordCountry` + `recordSubdivision` + `subjectId` | 2 |

26 total. It is latent today only because all 9 unscoped searches also omitted `subjectId`
— i.e. it goes live the moment #1073 lands.

**Both cases are one guard, and the decision is: suppress.** Unscoped and country-wide are
the same situation — a country-wide nil means the record is not in that country's indexed
collections, so naming counties inside it is noise. Fire only when the search was scoped to
something narrower than a country.

`placeTokens` collapses exactly the fact you need to test — its empty-fallback makes a
country-only place indistinguishable from a narrower one — so this wants `COUNTRY` hoisted
to module scope and a small exported predicate beside it, rather than the country set
duplicated into `record-search.ts`. Roughly:

```ts
// marriage-jurisdictions.ts
export function isSubCountryPlace(place: string | undefined): boolean { … }

// record-search.ts
if (isMarriageSearch && foundNobody && isSubCountryPlace(searchedPlace)
    && input.subjectId && input.projectPath) {
```

Worth two tests in the `describe` block and `TREE` fixture you already added here —
unscoped → no hint, `recordCountry` alone → no hint — plus a bullet in the spec's
"Fires when **all** of" list and a sentence in "Place matching".

### 2. Three comments now contradict the corrections made in this same PR

- `packages/engine/mcp-server/src/tools/record-search.ts:653` still says `subjectResolvable`
  is set *"only for a scoreable subject against a pool that genuinely lacks one"* — the exact
  claim the spec was corrected on, three lines of diff away at `record-search-tool-spec-v2.md:269`.
  The corrected wording is already there to copy.
- `packages/engine/mcp-server/src/types/record-search.ts:274-275` says *"Present only on a
  **NIL** marriage search"* (it also fires on `subjectResolvable: false`) and that candidates
  are *"earliest first"* — the ordering this entire revision replaced.
- `docs/specs/record-search-tool-spec-v2.md:288` — the table row reads *"Echo of the
  `marriagePlace` searched"*; it is now `marriagePlace` **or** `recordSubdivision` +
  `recordCountry`, or undefined. The prose section below it is correct.

Per CLAUDE.md the spec is what `spec-review` checks implementations against, so it is worth
not merging a diff that settles two questions in one file and re-asserts the old answers in
another.

### Not blocking: token containment is level-blind

```
tree:     "Texas, Missouri, United States"   (Texas County, Missouri)
searched: "Texas, United States"
result:   []
```

`ts.every(t => tc.includes(t))` ignores which level a token sits at, so a state name that is
also a county name elsewhere suppresses a legitimately different place. Not rare in US work —
Ohio County KY/IN/WV, Indiana County PA, Iowa County WI, Delaware County. It drops a lead
silently rather than adding a wrong one, so I would not hold the merge for it; a suffix test
instead of `every`/`includes` would close it, and I checked that all seven existing exclusion
tests still hold under one. A one-line caveat in the spec's "Place matching" section is
enough if you would rather not touch it here.

### Why no eval run

The guard can only ever remove output. And of the 26 marriage searches in the six runlogs,
the 6 that carried `subjectId` — the only ones the hint can reach at all — were every one of
them scoped narrower than a country:

```
run 07-30_14-32-18   marriagePlace = Hill County, Texas
run 07-30_14-32-18   marriagePlace = Hill County, Texas
run 07-30_23-05-46   marriagePlace = Hill County, Texas
run 07-30_23-05-46   recordSubdivision = Texas, recordCountry = United States
run 07-30_23-05-46   recordSubdivision = Texas, recordCountry = United States
run 07-31_13-02-13   marriagePlace = Hill County, Texas
```

So the guard changes behaviour on zero historical searches and the run-5/run-6 evidence is
untouched, exactly as with the last round. The right verifier is again a unit test built
from the real argument shapes.

Filed as follow-ups from this review: #1072 (general narrow-locality broadening, deliberately
outside `marriage-jurisdictions.ts`) and #1073 (`subjectId` coverage, which is what makes
finding 1 live).

File: docs/specs/record-search-tool-spec-v2.md (hunk 7 above)
Comment (DallanQ) [reply docs/specs/record-search-tool-spec-v2.md]: Spec side is fixed — `record-search-tool-spec-v2.md:269` now describes both branches and why the hint fires on each. Resolving.

One carry-over: the identical claim survives in the code comment at `packages/engine/mcp-server/src/tools/record-search.ts:653` ("the ranker sets **only** for a scoreable subject against a pool that genuinely lacks one"), so the spec and the comment now disagree. Picked up in the re-review rather than reopening this thread.

**Review by DallanQ — APPROVED** [review]
